### DISCUSSION TITLE
Best practice to export for IE (Internet Explorer) 11

### DISCUSSION THREAD

**Discussion by juacala** [babel/babel#13481] [Q&A] [answered]
I am trying to take various javascript files and modify them so they are IE11 compatible (no object dereferencing, arrow functions, etc.). 

I roughly followed this walkthrough among others: https://medium.com/@ramez.aijaz/transpile-typescript-to-es5-using-babel-and-webpack-f3b72a157399

Here's what I came up with:
package.json:
```
{
  "name": "espbabel",
  "version": "1.0.0",
  "description": "ESP Babel interface",
  "main": "index.js",
  "dependencies": {
    "core-js": "^3.14.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.14.5",
    "@babel/core": "^7.14.6",
    "@babel/plugin-proposal-decorators": "^7.14.5",
    "@babel/plugin-transform-runtime": "^7.12.17",
    "@babel/preset-env": "^7.12.17",
    "@babel/runtime": "^7.14.6",
    "babel-loader": "^8.2.2",
    "css-loader": "^5.0.2",
    "html-loader": "^2.1.0",
    "html-webpack-plugin": "^5.2.0",
    "sass": "^1.32.8",
    "style-loader": "^2.0.0",
    "webpack": "^5.23.0",
    "webpack-cli": "^4.5.0",
    "webpack-dev-server": "^3.11.2"
  },
  "scripts": {
    "build": "babel ../../../js -d ../../../js/ie"
  },
  "author": "",
  "license": "ISC"
}
```

config.bable.js:
```
module.exports = api => {
    api.cache(true)
    return {
        plugins: [
            ["@babel/plugin-proposal-decorators",{"decoratorsBeforeExport":true}],
            ["@babel/plugin-proposal-class-properties"],
            ["@babel/transform-runtime"]
        ],
/*      [
            "@babel/plugin-proposal-nullish-coalescing-operator",
            "@babel/plugin-proposal-optional-chaining",
            "@babel/plugin-transform-runtime",
        ],*/
        presets: [
            [
                "@babel/preset-env",
                {
                    modules: false,
                    useBuiltIns: "usage",
                    corejs: {
                        proposals: true,
                        version: 3
                    },
                    // caller.target will be the same as the target option from webpack
                    targets: {
                        // chrome: "58",
                        ie: "11"
                    }
                }
            ]
        ]
    }
```

among various other things. 

I run
`npm run build`
which converts the files. This all works fine, except for there are import or require statements at the top of the files that of course IE doesn't support. 

Is there a way to replace things like arrow functions/object dereferncing with their IE equivalents without getting the imports at the top of the file? 

What are others doing to convert codebases to be IE compliant? Apparently there is another step after babel runs to get it to actually work.

**Comment by nicolo-ribaudo** [comment, marked as answer]
You don't need `@babel/plugin-proposal-class-properties`, `@babel/plugin-proposal-nullish-coalescing-operator` and `@babel/plugin-proposal-optional-chaining`, since they are now included in `@babel/preset-env`.

> This all works fine, except for there are import or require statements at the top of the files that of course IE doesn't support.

You remove the `corejs`/`useBuiltIns` options, so that Babel doesn't inject imports for polyfills, and remove `@babel/transform-runtime`, so that Babel doesn't inject imports for runtime helpers.
However, you might then need to manually load a polyfill (either `core-js-bundle` or `es6-shim`, or something else).

**Reply by juacala** [reply]
That works great. See my answer below for exactly the config I ended up using.

**Comment by juacala** [comment]
Thank you very much! That does work.

Here is what I used in the end, and it works great.

babel.config.js:
```
module.exports = api => {
    api.cache(true)
    return {
	plugins: [
	    ["@babel/plugin-proposal-decorators",{"decoratorsBeforeExport":true}],
	],
	presets: [
	    [
		"@babel/preset-env",
		{
		    modules: false,
		    targets: {
			ie: "11"
		    }
		}
	    ]
	]
    }
}
```

package.json (although this is probably overkill by a lot):
```
{
    "name": "espbabel",
    "version": "1.0.0",
    "description": "ESP Babel interface",
    "private": true,
    "dependencies": {
	"core-js": "^3.14.0"
    },
    "devDependencies": {
	"@babel/cli": "^7.14.5",
	"@babel/core": "^7.14.6",
	"@babel/plugin-proposal-decorators": "^7.14.5",
	"@babel/plugin-transform-runtime": "^7.12.17",
	"@babel/preset-env": "^7.12.17",
	"@babel/runtime": "^7.14.6",
	"babel-loader": "^8.2.2",
	"css-loader": "^5.0.2",
	"html-loader": "^2.1.0",
	"html-webpack-plugin": "^5.2.0",
	"sass": "^1.32.8",
	"style-loader": "^2.0.0",
	"webpack": "^5.39.1",
	"webpack-cli": "^4.7.2",
	"webpack-dev-server": "^3.11.2"
    },
    "scripts": {
	"build": "babel ../../../js"
    },
    "author": "",
    "license": "ISC"
}
```

Then I run

```
npm run build -- -d [dest folder] [other options if you want]
```

and that duplicates the folder/file structure but in something that IE can understand. 

It isn't the most efficient since each file generates its own polyfills, but it suits my needs. I just need it to work. It being even slower than just because its IE hopefully will encourage our users to switch to a better browser, and we can still say we're IE compliant.

**Reply by nicolo-ribaudo** [reply]
If you don't want to duplicate helpers, you could switch back to your original config and use Webpack or Rollup to generate a single file resolving all the `import`/`require`s at compilation time.
